core/Pos/Crypto/SecretSharing.hs

```diff
@@ -120,77 +125,143 @@ instance Bi SecretProof =>
+    n = fromIntegral (length ps)
+    sorted = sort (toList ps)
+    convertRes (gen, secret, shares, comms, proof, pproofs) =
+        (coerce secret,
+         SecretProof gen proof pproofs comms,
+         zip sorted (coerce shares))
 
 -- | Recover secret if there are enough correct shares.
-recoverSecret :: Threshold -> [(EncShare, VssPublicKey, Share)] -> Maybe Secret
-recoverSecret t = recoverSecretDo t . getValidShares t
-
-recoverSecretDo :: Threshold -> [Share] -> Maybe Secret
-recoverSecretDo t shares
-    | genericLength shares >= t = Just (unsafeRecoverSecret shares)
-    | otherwise = Nothing
-
--- | Recover Secret from shares. Assumes that number of shares is
--- enough to do it. Consider using `getValidShares` first or
--- use `recoverSecret`
-unsafeRecoverSecret :: [Share] -> Secret
-unsafeRecoverSecret = Secret . Pvss.recover . map getShare
-
--- | Get #Threshold decrypted shares from given list, or less if there
--- is not enough.
-getValidShares :: Threshold -> [(EncShare, VssPublicKey, Share)] -> [Share]
-getValidShares t = map Share . Pvss.getValidRecoveryShares t . map convert
+--
+-- You *must* perform these checks on earlier stages:
+--
+--   * There are as many decrypted shares as there are encrypted shares for
+--     each participant.
+--
+--   * All shares are decrypted correctly (use 'verifyDecShare' for that)
+--
+recoverSecret
+    :: Scrape.Threshold
+    -> [(VssPublicKey, Int)]            -- ^ Participants + how many shares
+                                        --    were sent to each
+    -> HashMap VssPublicKey [DecShare]  -- ^ Shares decrypted and returned by
+                                        --    some participants
+    -> Maybe Secret
+recoverSecret (fromIntegral -> thr) (sortWith fst -> participants) shares = do
+    -- We reorder the shares so that 'recover' can consume them
+    let ordered :: [(Scrape.ShareId, DecShare)]
+        ordered = reorderDecryptedShares participants shares
+    -- Then we check that we have enough shares and do secret recovery
+    guard (length ordered >= thr)
+    pure (coerce Scrape.recover (take thr ordered))
+
+-- | Like 'Scrape.reorderDecryptedShares', but handles the case when multiple
+-- shares were created for a single key.
+--
+-- TODO: move to pvss-haskell, maybe
+reorderDecryptedShares
+    :: [(VssPublicKey, Int)]            -- ^ Participants
+    -> HashMap VssPublicKey [DecShare]  -- ^ Decrypted shares
+    -> [(Scrape.ShareId, DecShare)]
+reorderDecryptedShares participants shares =
+    map (first toInteger) $ go 1 participants
   where
-    convert (EncShare es, VssPublicKey p, Share s) = (es, p, s)
+    go :: Int                        -- ^ Index of current share
+       -> [(VssPublicKey, Int)]      -- ^ Remaining participants
+       -> [(Int, DecShare)]
+    go _ []         = []
+    go i ((k,n):ps) = case HM.lookup k shares of
+        Nothing -> go (i + n) ps
+        Just ss -> zip [i..] (take n ss) ++ go (i + n) ps
```

Comment:
```haskell
-- __Description of the algorithm__
--
-- We know:
--   * the /original/ order of participants
--   * how many shares were generated for each participant
--   * a list of decrypted shares for each participant,
--     though some participants might be missing
--
-- /Note:/ we assume that if a participant isn't missing then their shares
-- are present in the right order and no shares are skipped. This is a valid
-- assumption to make because
--   * first we verify encrypted shares with 'verifyEncShares'
--     (this takes care of the order and count of shares)
--   * then we verify each decrypted share with 'verifyDecShare' against
--     the corresponding encrypted share
--   * and we don't forget to check that the counts of encrypted and
--     decrypted shares match.
--
-- After this is done, we 'go' through the list of participants and try to
-- recover a share with index 'i' for each 'i'. If we have the shares for
-- some participant (let's say 'n' shares), we can recover shares with
-- indices @[i..i+n-1]@, so we add them to the list and continue from index
-- @i+n@. If we don't have the shares for the participant, we just skip 'n'
-- shares and move to the next participant and try to find the @i+n@'th
-- share.
reorderDecryptedShares
    :: [(VssPublicKey, Int)]            -- ^ Participants + how many shares
                                        --    were sent to each
    -> HashMap VssPublicKey [DecShare]  -- ^ Decrypted shares
    -> [(Scrape.ShareId, DecShare)]
reorderDecryptedShares participants shares =
    map (first toInteger) $ go 1 participants
  where
    go :: Int                        -- ^ Index of current share
       -> [(VssPublicKey, Int)]      -- ^ Remaining participants
       -> [(Int, DecShare)]
    go _ []         = []
    go i ((k,n):ps) = case HM.lookup k shares of
        Nothing -> go (i + n) ps
        Just ss -> zip [i..] (take n ss) ++ go (i + n) ps
```

I have added an explanation, please review it.